typescript/src/proof.ts

```diff
@@ -70,3 +79,253 @@ function createMerkleProof(txMerkleBranch: TransactionMerkleBranch): string {
+) {
+  if (requiredConfirmations < 1) {
+    throw new Error("The number of required confirmations but at least 1")
+  }
+
+  const proof = await assembleTransactionProof(
+    transactionHash,
+    requiredConfirmations,
+    bitcoinClient
+  )
+
+  const bitcoinHeaders: BlockHeader[] = splitHeaders(proof.bitcoinHeaders)
+  if (bitcoinHeaders.length != requiredConfirmations) {
+    throw new Error("Wrong number of confirmations")
+  }
+  const merkleRootHash: Hex = bitcoinHeaders[0].merkleRootHash
+
+  validateMerkleTree(
+    transactionHash.reverse().toString(),
+    merkleRootHash.toString(),
+    proof.merkleProof,
+    proof.txIndexInBlock
+  )
+
+  validateBlockHeadersChain(
+    bitcoinHeaders,
+    previousDifficulty,
+    currentDifficulty
+  )
+}
+
+/**
+ * Validates the Merkle tree by checking if the provided transaction hash,
+ * Merkle root hash, intermediate node hashes, and transaction index parameters
+ * produce a valid Merkle proof.
+ * @param transactionHash The hash of the transaction being validated.
+ * @param merkleRootHash The Merkle root hash that the intermediate node hashes
+ *        should compute to.
+ * @param intermediateNodeHashes The Merkle tree intermediate node hashes,
+ *        concatenated as a single string.
+ * @param transactionIndex The index of the transaction being validated within
+ *        the block, used to determine the path to traverse in the Merkle tree.
+ * @throws {Error} If the Merkle tree is not valid.
+ * @returns An empty return value.
+ */
+function validateMerkleTree(
+  transactionHash: string,
+  merkleRootHash: string,
+  intermediateNodeHashes: string,
+  transactionIndex: number
+) {
+  // Shortcut the empty-block case
+  if (
+    transactionHash == merkleRootHash &&
+    transactionIndex == 0 &&
+    intermediateNodeHashes.length == 0
+  ) {
+    return
+  }
+
+  validateMerkleTreeHashes(
+    transactionHash,
+    intermediateNodeHashes,
+    merkleRootHash,
+    transactionIndex
+  )
+}
+
+/**
+ * Validates the transaction's Merkle proof by traversing the Merkle tree
+ * starting from the provided transaction hash and using the intermediate node
+ * hashes to compute the root hash. If the computed root hash does not match the
+ * merkle root hash, an error is thrown.
+ * @param transactionHash The hash of the transaction being validated.
+ * @param intermediateNodesHashes The Merkle tree intermediate nodes hashes,
+ *        concatenated as a single string.
+ * @param merkleRootHash The Merkle root hash that the intermediate nodes should
+ *        compute to.
+ * @param transactionIndex The index of the transaction in the block, used
+ *        to determine the path to traverse in the Merkle tree.
+ * @throws {Error} If the intermediate nodes are of an invalid length or if the
+ *         computed root hash does not match the merkle root hash parameter.
+ * @returns An empty return value.
+ */
+function validateMerkleTreeHashes(
+  transactionHash: string,
```

Comment:
We can do
```suggestion
  transactionHash: TransactionHash,
```
and reverse the hash within this function, right in the place of use.

Reply:
Done in 57c55f4e3fc5cb0446dafeb566e274134ba9d77c.